### DISCUSSION TITLE
How to avoid leaving needless things when building macOS app in single-dir mode?

### DISCUSSION THREAD

**Discussion by mhucka** [pyinstaller/pyinstaller#6562] [Help] [answered]
I finally figured out the right configuration and file structure to create macOS `.app` bundles in single-dir mode. However, my PyInstaller configuration is still resulting in the creation of a single-file app and a folder (probably from the one-dir setting), even though I don't need them. I don't see how to avoid this because of the way the PyInstaller configuration values build on each other, but maybe I still don't understand how to do it best. Here's the relevant part of the code (full config is [here](https://github.com/caltechlibrary/foliage/blob/main/pyinstaller-macos.spec)):

```python
executable         = EXE(application_pyz,
                         configuration.scripts,
                         configuration.binaries,
                         configuration.zipfiles,
                         configuration.datas,
                         name = 'Foliageapp',
                         icon = r'dev/icon/foliage-icon.icns',
                         strip = False,
                         upx = True,
                         runtime_tmpdir = None,
                         bootloader_ignore_signals = False,
                         codesign_identity = None,
                         entitlements_file = None,
                         debug = False,
                         console = True,
                        )

collected_files = COLLECT(executable,
                          configuration.binaries,
                          configuration.zipfiles,
                          configuration.datas,
                          strip = False,
                          upx = False,
                          upx_exclude = [],
                          name = 'foliage')

app             = BUNDLE(collected_files,
                         name = 'Foliage.app',
                         icon = 'dev/icon/foliage-icon.icns',
                         version = version,
                         bundle_identifier = 'edu.caltech.library.foliage',
                         info_plist = {'NSHighResolutionCapable': 'True',
                                       'NSPrincipalClass': 'NSApplication',
                                       'NSAppleScriptEnabled': False},
                        )
```

The result is 3 things instead of just the `Foliage.app` bundle that I am interested in:

```txt
dist/macos/
├── Foliage.app/
├── Foliageapp
└── foliage/
```

Is there a way to avoid the creation of `Foliageapp` and the `foliage` directory, or do people handle this by just deleting them after PyInstaller is done?

**Comment by rokm** [comment, marked as answer]
This looks incorrect to me. The `EXE()` entry and its options look like they came from a `onefile` preset, and the `COLLECT()` entry and its options look like they came from a `onedir` preset, and then they were manually stitched together. 

That's why you get a `onefile`-style executable (`Foliageapp` file), a `onedir`-style directory (`foliage`) that probably has an executable that actually behaves like a `onefile` execitable (and all extra collected stuff is redundant), and an app bundle that probably also has a onefile-style executable plus redundant collected stuff. Typically, you should be getting a regular onefile or onedir build, plus an app bundle (and then you distribute the one you want, in this case probably the app bundle).

Try generating a new ("clean") spec file for onefile or onedir build (whichever you actually want) with windowed option enabled, and then work from that. Or take a look [here for an example of switchable .spec file](https://github.com/pyinstaller/pyinstaller/blob/develop/tests/functional/specs/pyi_osx_aevent_handling_carbon.spec) as an illustration of differences between onefile and onedir build (i.e., corresponding entries and options passed to them).

**Reply by mhucka** [reply]
Thanks for trying to help, but that's not correct; while I _did_ edit the conf file manually, a careful comparison between what I'm using and the [above-mentioned example](https://github.com/pyinstaller/pyinstaller/blob/develop/tests/functional/specs/pyi_osx_aevent_handling_carbon.spec) shows they are structurally the same: `BUNDLE` references `COLLECT` references `EXE` references `PYZ` references `Analysis`, and while my `EXE` has some arguments that aren't in the example, they are necessary to make the app work (something that I can demonstrate by removing them and testing if the result runs).

<s>But I found the problem, and it turns out it's not because of the conf file. Rather embarrassingly, my makefile's rule for cleaning past products was faulty, and what I was seeing was ... left over from prior builds in one-file mode :-(.</s>

<s>Thanks again, and sorry for the goose chase!</s>

**Reply by rokm** [reply]
> shows they are structurally the same

In onedir mode, `EXE()` receives `exclude_binaries=True` and does not receive `a.binaries`, `a.zipfiles`, `a.datas`, because those are collected by `COLLECT()`. 

So onedir mode is `EXE without binaries & data` + `COLLECT` + `BUNDLE`, while onefile mode is `EXE` + `BUNDLE`. What you have here is a mixture of both.

**Reply by mhucka** [reply]
You're right, and now that I tried changing it the way you describe, re-read your original explanation, I now understand that getting a one-dir output _in addition to_ the `.app` is actually expected. So now I get

```txt
dist/macos/
├── Foliage.app/
└── foliage/
```

My apologies. Serves me right for working too quickly ...
